### PR TITLE
Update to Diesel 2.0 (rc0)

### PR DESCRIPTION
**Pull request by felinira**
This updates / basically rewrites the code to work with diesel 2.0 rc0. Maybe it is useful for someone. As there can be any number of changes to diesel in the meantime this might break at any time.

### CODE REVIEW

**Comment by BlackDex** [comment]
@felinira this doesn't seem to work with Diesel `v2.0.0-rc0`.

**Comment by felinira** [comment]
@BlackDex I updated the code but didn't yet have a chance to test it. If you use this I would appreciate feedback if everything works

**Comment by BlackDex** [comment]
@felinira, i have some time tomorrow to test this i think.
I will let you know how that goes!

**Comment by BlackDex** [comment]
@felinira I just did a quick check, but trying to enable this on Diesel v2 causes the following error(s):

```
the trait bound `SqliteConnection: diesel::connection::Connection` is not satisfied
the following other types implement trait `diesel::connection::Connection`:
  LoggingConnection<C>
  PooledConnection<M>
```

```
the trait bound `LoggingConnection<SqliteConnection>: R2D2Connection` is not satisfied
the trait `R2D2Connection` is implemented for `SqliteConnection`
required because of the requirements on the impl of `ManageConnection` for `ConnectionManager<LoggingConnection<SqliteConnection>>`
```

```
the method `get` exists for struct `Pool<ConnectionManager<LoggingConnection<SqliteConnection>>>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`ConnectionManager<LoggingConnection<SqliteConnection>>: ManageConnection`
```

```
the trait bound `LoggingConnection<_>: Connection` is not satisfied
the following other types implement trait `Connection`:
  PooledConnection<M>
  SqliteConnection
```

<details><summary>Using this code btw:</summary>

```rust
#[macro_use]
extern crate diesel;
// Also include diesel_migrations because it causes some other issues during compiling.
#[macro_use]
#[allow(unused_imports)]
extern crate diesel_migrations;

#[macro_use]
extern crate log;

use diesel::r2d2::{ConnectionManager, Pool};

#[cfg(feature = "logging")]
use diesel_logger::LoggingConnection;


mod schema {
    table! {
        posts (id) {
            id -> Int4,
            title -> Varchar,
            body -> Text,
            published -> Bool,
        }
    }
}

mod models {
    use crate::schema::posts;
    #[derive(Queryable, Identifiable, Insertable, AsChangeset)]
    pub struct Post {
        pub id: i32,
        pub title: String,
        pub body: String,
        pub published: bool,
    }

    // apparently this can be done without heap storage, but lifetimes spread far..
    #[derive(Insertable)]
    #[diesel(table_name = posts)]
    // #[table_name = "posts"]
    pub struct NewPost {
        pub title: String,
        pub body: String,
    }
}

use diesel::{prelude::*, insert_into};
use diesel::sqlite::SqliteConnection;

#[cfg(feature = "logging")]
pub type DbConnection = LoggingConnection<SqliteConnection>;

#[cfg(not(feature = "logging"))]
pub type DbConnection = SqliteConnection;

fn main() {
    use schema::posts::dsl::*;
    env_logger::init();

    info!("Start");

    let database_url = std::env::var("DATABASE_URL").unwrap_or_else(|_| "main.db".into());

    let manager = ConnectionManager::<DbConnection>::new(&database_url);
    let pool = Pool::builder().max_size(1).build(manager).unwrap();

    #[cfg(not(feature = "logging"))]
    let mut conn = pool.get().unwrap();

    #[cfg(not(feature = "logging"))]
    let conn = &mut conn;

    #[cfg(feature = "logging")]
    let conn = pool.get().unwrap();

    #[cfg(feature = "logging")]
    let conn = &mut LoggingConnection::new(conn);

    insert_into(posts).values(title.eq("Test1")).execute(conn).unwrap();

    info!("End");
}
```

</details>

**Comment by felinira** [comment]
Ah yeah, I still had the git repo configured. After changing to 2.0.0-rc.0 from crates.io your example works for me.

**Comment by BlackDex** [comment]
Yea, that works like a charm!

**Review by BlackDex — APPROVED** [review]

File: src/lib.rs [hunk 1]
```diff
@@ -82,22 +92,25 @@ where
...
-        T::Query: QueryFragment<Self::Backend> + QueryId,
+        T: 'query + Query + QueryFragment<Self::Backend> + QueryId,
         Self::Backend: QueryMetadata<<T as AsQuery>::SqlType>,
     {
         // We need to use unsafe copying here bc load takes the source by value
-        let debug_query = unsafe {
-            let tmp = ptr::read_unaligned(&source) ;
-            debug_query::<<LoggingConnection<C> as Connection>::Backend, _>(&tmp.as_query()).to_string()
-        };
+        let tmp = unsafe { ptr::read_unaligned(&source) };
```
Comment (SadiinsoSnowfall) [inline src/lib.rs]: This is causing one of my projects to segfault. Are you sure this is sound ? Have you tested it against Miri ?

Comment (felinira) [reply src/lib.rs]: It's not and I am struggling to find out why I thought this was the only way to do it. Fixed it, sorry.

Comment (SadiinsoSnowfall) [reply src/lib.rs]: Thank you for the quick fix, I tried to find how to format the query on my own but couldn't manage to do it, I was stuck at trying to get a Backend instance to pass it to `to_sql`.

Comment (felinira) [reply src/lib.rs]: This whole thing is kind of a mess and a big hack. But at least it doesn't use unsafe code now.

Comment (SadiinsoSnowfall) [reply src/lib.rs]: Small nitpick: 
you can directly do
```rust
let debug_string = debug_query(&query).to_string();
```
Template type inference should work, and no need to use the format macro just for `to_string`

**Comment by shssoichiro** [comment]
Sorry I've taken so long to review this. Although it seems like diesel 2.0 is still not out of rc either :sweat_smile:

**Pull request merged**